tutorials/graph-theory/binary-search-tree.md

```diff
@@ -1,12 +1,304 @@
+
+![image](https://user-images.githubusercontent.com/30936607/195677540-23c39b0f-ce06-4c7d-bffa-468b6c1b3f79.png)
+
+Like binary tree. We can also traverse BST in **pre**,**post** and **inorder** traversal.
+But the amazing fact is that **inorder** traversal
+of `BST` will give **sorted data**.
+
+### Inorder traversal in BST
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+#include<bits/stdc++.h>
+using namespace std;
+
+class Node{
+  public:
+    int data;
+    Node* left;
+    Node* right;
+    Node(int d){
+        data = d;
+        left = NULL;
+        right = NULL;
+     }
+};
+
+void inOrder(Node *root){
+  if(root==NULL)
+    return;
+  inOrder(root->left);
+  //output: 4 6 7 12 13 15 16
+  cout<<root->data<<" ";
+  inOrder(root->right);
+}
+
+int main(){
+  Node *root = new Node(12);
+  root->left = new Node(6);
+  root->left->left = new Node(4);
+  root->left->right = new Node(7);
+  root->right = new Node(15);
+  root->right->left = new Node(13);
+  root->right->right = new Node(16);
+  inorder(root);
+  return 0;
+}
+
+```
+
+</TabItem>
+
+</Tabs>
+
+### Operations on BST
+
+There are three main operation in BST:
+
+- Search
+- Insertion
+- Deletion
+
+### 1. Search
+
+##### Algorithm
+
+1. The searching always start from the root node of the tree.
+2. We compare the data in each node with our target node that we want to search in tree.
+3. If our target node value matches with the value of current node then return that node.
+4. If value not matched with current node then we either procced to right or left child according to the condition.
+5. If target node value is greater than the current node then we procced towards the right child or right subtree.
+6. If target node value is lower than the current node then we proceed towards the left child or left subtree.
+
+`target value: value of node you want to search in tree`
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+// For BST search
+Node *bstSearch(int target,Node* root){
+    if(root == NULL)
+        return root;
+    if(root->data == target)
+        return root;
+    if(root->data<target){
+      // search target node in right subtree
+        return bstSearch(target,root->right);
+    } else{
+       // search target node in left subtree
+        return bstSearch(target,root->left);
+    }
+}
+```
+
+</TabItem>
+</Tabs>
+
+### 2. Insertion
+
+**Algorithm**
+
+1. Insertion begin with searh operation.
+2. Search left or right subtree according to the relation between node value and the value of our target node as we did in search.
+3. Repeat step 1 until we reach to correct place to insert our new node.
+4. Add a new node as a right or left child according to the node value and our target node value.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
```

Comment:
Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example